pkg/skaffold/docker/image.go

```diff
@@ -312,13 +312,18 @@ func (l *localDaemon) isAlreadyPushed(ctx context.Context, ref, registryAuth str
 
 // Pull pulls an image reference from a registry.
 func (l *localDaemon) Pull(ctx context.Context, out io.Writer, ref string) error {
+	// Eargerly create credentials.
 	registryAuth, err := l.encodedRegistryAuth(ctx, DefaultAuthHelper, ref)
-	if err != nil {
-		return fmt.Errorf("getting auth config for %q: %w", ref, err)
-	}
+	// Let's ignore the error because maybe the image is public
+	// and can be pulled without credentials.
 
 	rc, err := l.apiClient.ImagePull(ctx, ref, types.ImagePullOptions{
 		RegistryAuth: registryAuth,
+		PrivilegeFunc: func() (string, error) {
+			// If the first pull is unauthorized, retry without ignoring the
+			// authentication error.
+			return registryAuth, err
```

Comment:
So AFAICT this doesn't *retry*: 
  1. If `encodedRegistryAuth()` returns an error, this anonymous function returns that error causing `ImagePull()` to short-circuit and immediately return that error — it doesn't retry. 
  2. If `encodedRegistryAuth()` had been able to obtain `registryAuth` without an error, then `ImagePull()` will just retry with the auth and get the same unauthorized error.

If `encodedRegistryAuth()` did encounter an error, then `registryAuth` is an empty string.  Maybe it would make sense have this anonymous function instead do `return "", err`?  So that
  - if `encodedRegistryAuth()` had encountered an error, then we've tried the no-auth case (since `registryAuth == ""`, and so we use `ImagePull()`'s short-circuiting to return the auth error
  - if `encodedRegistryAuth()` didn't encounter an error (so `err == nil`), but our credential was seeming invalid (unauthorized), then we'll retry with no auth

```suggestion
			// The first pull is unauthorized. There are two situations:
			//   1. if `encodedRegistryAuth()` errored, then `registryAuth == ""` and so we've
			//     tried an anonymous pull which has failed.  So return the original error from
			//     `encodedRegistryAuth()`.
			//   2. If `encodedRegistryAuth()` succeeded (so `err == nil`), then our credential was rejected, so
			//     return "" to retry as an anonymous pull.
			return "", err
```

Reply:
Alternatively we could invert the logic of this function and try a pull with no auth first, and then retrying with auth if that fails?  It's an additional round-trip, and counter to what Docker does.  Though it is the process Jib uses.

Reply:
This may be complete irrelevant, but just FYI, GCR and Docker Hub have different behaviors for public images.

- GCR: public images in are completely public. No strings attached.
- Docker Hub: it always returns "401 Unauthorized" first, and it wants you to always go through the bearer auth token authentication. Of course, for public images, you don't have to send any credentials when talking to `auth.docker.io/token` to get a token. I don't know if go-containerregistry or whatever in Skaffold is automatically taking care of this under the hood though. Here's an example when accessing `openjdk` (i.e., `GET https://registry-1.docker.io/v2/library/openjdk/manifests/8`).
```
CONFIG: -------------- RESPONSE --------------
HTTP/1.1 401 Unauthorized
...
Www-Authenticate: Bearer realm="https://auth.docker.io/token",service="registry.docker.io",scope="repository:library/openjdk:pull"
```